### PR TITLE
[mono] remove 2 extra members from System.Object

### LINKED ISSUE #15029: [netcore] Make System.Reflection.Tests.TypeInfoTests.FindMembers Pass
The test currently fails with:

```
System.Reflection.Tests.TypeInfoTests.FindMembers [FAIL]
      Assert.Equal() Failure
      Expected: 28
      Actual:   30
      Stack Trace:
           at System.Reflection.Tests.TypeInfoTests.FindMembers()
           at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

It appears we have two extra instance methods on Object, which makes this test fail.  If that can't change, we may want to modify the tests to detect which VM is running and act accordingly.

### PR DESCRIPTION
**Pull request by lambdageek**
1. move `GetRawData` to an extension method in RuntimeHelpers and fixup the intrinsics
2. replace `Object.CloneInternal` by `MarshalAsAttribute.CloneInternal` at its one use site
3. re-enable System.Reflection.Tests.TypeInfoTests.FindMembers and fix mono/mono#15029

### CODE REVIEW

**Comment by CoffeeFlux** [comment]
Relevant: https://github.com/mono/mono/pull/17295

**Comment by CoffeeFlux** [comment]
Also related: https://github.com/mono/mono/issues/15029

**Comment by vargaz** [comment]
This might depend on optimizations that are missing in the mono JIT/interpreter to get the same performance.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 1]
```diff
@@ -132,5 +132,14 @@ public static void PrepareConstrainedRegionsNoOP()
         internal static bool IsPrimitiveType(this CorElementType et)
             // COR_ELEMENT_TYPE_I1,I2,I4,I8,U1,U2,U4,U8,R4,R8,I,U,CHAR,BOOLEAN
             => ((1 << (int)et) & 0b_0011_0000_0000_0011_1111_1111_1100) != 0;
+
+        internal static ref byte GetRawData(this object obj) =>
+            ref Unsafe.As<RawData>(obj).Data;
+    }
+    // Helper class to assist with unsafe pinning of arbitrary objects.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs:139]: Nit:
```suggestion
    }

    // Helper class to assist with unsafe pinning of arbitrary objects.
```

**Review by stephentoub — DISMISSED** [review]
Yay sharing more stuff.

**Comment by CoffeeFlux** [comment]
Filip seemed to think the codegen ended up being the same, but I'd like to be sure of that before we merge this.

**Review by BrzVlad — APPROVED** [review]

**Comment by lambdageek** [comment]
Zoltan:
> This might depend on optimizations that are missing in the mono JIT/interpreter to get the same performance.
Ryan:
> Filip seemed to think the codegen ended up being the same, but I'd like to be sure of that before we merge this.

So I think on the JIT this is ok - there's an extra null check compared to the intrinsic, but as Filip mentioned it would get eliminated if the extension method is inlined.

On the other hand, in  the interpreter the null check isn't done in the transform, it's done in the interpretation of `MINT_LDFLDA`, so it seems like there's an extra cost here.

---

So maybe this is a bad idea after all.  One thing we can do is move the method to be an extension method in `RuntimeHelpers.Mono.cs`, but leave it as an intrinsic.  We wouldn't get as much sharing with CoreCLR, but at least we'd do something about mono/mono#15029

**Comment by BrzVlad** [comment]
@lambdageek so the interpreter only generated a MINT_LDFLDA instruction for this ?

**Comment by vargaz** [comment]
Imho, moving these small pieces of low level code to shared is problematic in general, they do depend on implementation details in the runtime.

**Comment by lambdageek** [comment]
Ok, I'm going to retool this PR to:
1. move `GetRawData` to an extension method in RuntimeHelpers and fixup the intrinsics
2. replace `CloneInternal` by `MemberwiseClone` at its one use site
3. re-enable `System.Reflection.Tests.TypeInfoTests.FindMembers` and fix https://github.com/mono/mono/issues/15029

**Review by CoffeeFlux — APPROVED** [review]
Much happier with this approach, thanks!

LGTM, but hopefully Vlad can take a look at the interpreter changes.

**Comment by BrzVlad** [comment]
@lambdageek now that you brought the idea up, you could use MINT_LDFLDA_UNSAFE instead on the interpreter for better perf. I can add it in a separate PR though, if you feel it's unrelated.

**Comment by lambdageek** [comment]
I think I'm ok with the PR as is.

I think Zoltan makes a compelling argument that we shouldn't try to share low-level hacks like the `Unsafe.As<RawData>` implementation of `GetRawData`.

**Pull request merged**